python/graphstorm/graphpartitioning/metadata_schema.py

```diff
@@ -0,0 +1,730 @@
+        edge of the input graph. These ids are expected to be unique ids
+        across the input graph.
+
+        The range of global_edge_ids for each edge type in the graph are
+        stored in a dictionary where key-value pairs are etypes and a tuple
+        which indicates the starting and ending value (exlcusive) of the
+        global_edge_ids. For instance,
+            etype1: (a, b) where
+            etype1 is a string describing the edge type
+            and [a, b) is the range of global_edge_ids for the edge type
+            `etype1`.
+        """
+        etypes = self._data[STR_EDGE_TYPE]
+        etype_counts = self._data[STR_NUM_EDGES_PER_TYPE]
+        for count in etype_counts:
+            if isinstance(count, int) and count > 0:
+                continue
+            raise ValueError("Invalid value for No. of nodes per type")
+
+        prefix_sum = np.cumsum([0] + etype_counts)
+        starts = prefix_sum[:-1]
+        ends = prefix_sum[1:]
+        ranges = zip(starts, ends)
+
+        self._global_eid_offsets = dict(zip(etypes, ranges))
+
+    def _init_global_nid_offsets(self):
+        """Function to initialize the global_node_id offsets for all the
+        node types present in the input graph.
+
+        global_node_ids, or global_nids, are unique ids assigned to each
+        node of the input graph. These ids are expected to be unique ids
+        across the input graph.
+
+        The range of global_node_ids for each node type in the graph are
+        stored in a dictionary where key-value pairs are ntypes and a tuple
+        which indicates the starting and ending value (exclusive) of the
+        global_node_ids. For instance,
+            ntype1: [a, b) are the key-value pairs.
+            where ntype1 is a string describing the node type
+            and [a, b) is the range of global_node_ids for the node type
+            `ntype1`.
+        """
+        ntypes = self._data[STR_NODE_TYPE]
+        ntype_counts = self._data[STR_NUM_NODES_PER_TYPE]
+        for count in ntype_counts:
+            if isinstance(count, int) and count > 0:
+                continue
+            raise ValueError("Invalid value for No. of nodes per type")
+
+        prefix_sum = np.cumsum([0] + ntype_counts)
+        starts = prefix_sum[:-1]
+        ends = prefix_sum[1:]
+        ranges = zip(starts, ends)
+        self._global_nid_offsets = dict(zip(ntypes, ranges))
+
+    def _init_etype_maps(self):
+        """Initialization function to build unique id maps for node types of
+        the input graph. This function creates a list of edge types present in
+        the graph, and unique-id <-> edge type and reverse maps.
+        """
+        self._etypes = self._data[STR_EDGE_TYPE]
+        self._etype_id_map = {
+            etype_name: idx for idx, etype_name in enumerate(self._etypes)
+        }
+        self._id_etype_map = dict(enumerate(self._etypes))
```

Comment:
why do you need this? can't we use `self._etypes` ?

Reply:
The idea of creating a map between etype and etype_id and vice versa is to decouple the user to do so. This is the only one place which assigns ids to etypes. And if there is a logic change then there is only one place to change. This also serves to abstract away the logic from the application, which should be agnostic about how the ids are assigned to the etypes in the input graph.

Reply:
how is `self._id_etype_map` different from `self._etypes`?

Reply:
`self._id_etype_map[i] == self._etypes[i]`